File: example/src/main/java/ru/virgil/example/box/BoxController.java
package ru.virgil.example.box;

import com.google.common.truth.Truth;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.prepost.PostAuthorize;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import ru.virgil.example.system.HttpAddressConstants;
import ru.virgil.example.truck.Truck;
import ru.virgil.example.truck.TruckService;
import ru.virgil.example.user.UserDetails;
import ru.virgil.example.user.UserDetailsService;

import javax.annotation.security.RolesAllowed;
import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/box")
@RequiredArgsConstructor
public class BoxController implements HttpAddressConstants {

    private final UserDetailsService userDetailsService;
    private final TruckService truckService;
    private final BoxService boxService;
    private final BoxMapper boxMapper;

    @GetMapping
    public List<BoxDto> getAll(@RequestParam(PAGE_PARAM) int page, @RequestParam(PAGE_SIZE_PARAM) int size) {
        UserDetails owner = userDetailsService.getCurrentUser();
        return boxService.getAll(owner, page, size).stream()
                .map(boxMapper::toDto)
                .toList();
    }

    @GetMapping("/{uuid}")
    @PostAuthorize("""
            hasRole('ROLE_POLICE') and @boxSecurity.hasWeapon(returnObject)
            or hasRole('ROLE_USER') and not @boxSecurity.hasWeapon(returnObject)
            """)
    public BoxDto get(@PathVariable UUID uuid) {
        UserDetails owner = userDetailsService.getCurrentUser();
        Box box = boxService.get(owner, uuid);
        return boxMapper.toDto(box);
    }

    @PostMapping
    @PreAuthorize("""
            hasRole('ROLE_POLICE') and @boxSecurity.hasWeapon(#boxDto)
            or hasRole('ROLE_USER') and not @boxSecurity.hasWeapon(#boxDto)
            """)
    public BoxDto post(@RequestBody BoxDto boxDto) {
        Truth.assertThat(boxDto.getType()).isNotNull();
        UserDetails owner = userDetailsService.getCurrentUser();
        Truck assignedTruck = truckService.assignTruck();
        Box box = boxMapper.toEntity(boxDto);
        box = boxService.create(owner, assignedTruck, box);
        return boxMapper.toDto(box);
    }

    @PutMapping("/{uuid}")
    public BoxDto put(@PathVariable UUID uuid, @RequestBody BoxDto patch) {
        UserDetails owner = userDetailsService.getCurrentUser();
        Box patchBox = boxMapper.toEntity(patch);
        Box serverBox = boxService.edit(owner, uuid, patchBox);
        return boxMapper.toDto(serverBox);
    }

    @DeleteMapping("/{uuid}")
    public void delete(@PathVariable UUID uuid) {
        UserDetails owner = userDetailsService.getCurrentUser();
        boxService.delete(owner, uuid);
    }

    @GetMapping("/weapons")
    @RolesAllowed("ROLE_POLICE")
    public List<BoxDto> getAllWeapons() {
        UserDetails owner = userDetailsService.getCurrentUser();
        return boxService.getAllWeapons(owner).stream()
                .map(boxMapper::toDto)
                .toList();
    }
}

File: example/src/main/java/ru/virgil/example/box/BoxController.kt
package ru.virgil.example.box

import org.springframework.security.access.prepost.PostAuthorize
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import ru.virgil.example.system.rest.RestValues
import ru.virgil.example.truck.TruckService
import java.util.*
import javax.annotation.security.RolesAllowed

@RestController
@RequestMapping("/box")
class BoxController(
    private val truckService: TruckService,
    private val boxService: BoxService,
    private val boxSecurity: BoxSecurity,
) : BoxMapper {

    @GetMapping
    fun getAll(
        @RequestParam(RestValues.PAGE_PARAM) page: Int, @RequestParam(RestValues.PAGE_SIZE_PARAM) size: Int,
    ): List<BoxDto> = boxService.getAll(page, size).stream()
        .map { it.toDto() }
        .toList()

    @RolesAllowed("ROLE_POLICE")
    @GetMapping("/weapons")
    fun getAllWeapons(): List<BoxDto> = boxService.getAllMyWeapons().stream()
        .map { it.toDto() }
        .toList()

    @GetMapping("/{uuid}")
    @PostAuthorize(
        """
        hasRole('ROLE_POLICE') and @boxSecurity.hasWeapon(returnObject)
        or hasRole('ROLE_USER') and not @boxSecurity.hasWeapon(returnObject)
        """
    )
    operator fun get(@PathVariable uuid: UUID): BoxDto {
        val box = boxService.get(uuid)
        return box.toDto()
    }

    @PostMapping
    @PreAuthorize(
        """
        hasRole('ROLE_POLICE') and @boxSecurity.hasWeapon(#boxDto)
        or hasRole('ROLE_USER') and not @boxSecurity.hasWeapon(#boxDto)
        """
    )
    fun post(@RequestBody boxDto: BoxDto): BoxDto {
        // TODO: Заменить на Object Provider
        val assignedTruck = truckService.assignTruck()
        val createdBox = boxService.create(assignedTruck, boxDto)
        return createdBox.toDto()
    }

    @PutMapping("/{uuid}")
    fun put(@PathVariable uuid: UUID, @RequestBody boxDto: BoxDto): BoxDto {
        val serverBox = boxService.edit(uuid, boxDto)
        return serverBox.toDto()
    }

    @DeleteMapping("/{uuid}")
    fun delete(@PathVariable uuid: UUID) {
        boxService.delete(uuid)
    }
}
